File: Binary Tree Paths/Binary_Tree_Paths.py
# Runtime: 52 ms, faster than 18.21% of Python3 online submissions for Binary Tree Paths.
# Memory Usage: 13.3 MB, less than 5.44% of Python3 online submissions for Binary Tree Paths.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        res = []
        vec = []
        resint = []

        myStack = []
        preNode = None

        while root or myStack:
            while root:
                myStack.append(root)
                vec.append(root.val)
                root = root.left
            if myStack:
                root = myStack.pop()
                if root.right is None or root.right == preNode:
                    if root.left is None and root.right is None:
                        resint.append(vec[:])
                    vec.pop()

                    preNode = root
                    root = None
                else:
                    myStack.append(root)
                    root = root.right
            
        self.toString(resint, res)
        return res
    
    def toString(self, resint, res):
        for path in resint:
            res.append('->'.join([str(item) for item in path]))

File: Binary Tree Paths/Binary_Tree_Paths.cpp
// 二叉树的后序遍历

// Runtime: 4 ms, faster than 100.00% of C++ online submissions for Binary Tree Paths.
// Memory Usage: 10.2 MB, less than 100.00% of C++ online submissions for Binary Tree Paths.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution 
{
public:
    vector<string> binaryTreePaths(TreeNode* root) 
    {
        vector<string> res;
        vector<int> vec;
        vector<vector<int>> resint;

        stack<TreeNode*> myStack;
        TreeNode* preNode = nullptr;

        while (root || !myStack.empty())
        {
            while (root)
            {
                myStack.push(root);
                vec.push_back(root->val);
                root = root->left;
            }
            if (!myStack.empty())
            {
                root = myStack.top();
                myStack.pop();

                if (root->right == nullptr || root->right == preNode)
                {
                    if (root->left == nullptr && root->right == nullptr)
                        resint.push_back(vec);
                    vec.pop_back();

                    preNode = root;
                    root = nullptr;
                }
                else
                {
                    myStack.push(root);
                    root = root->right;
                }
            }
        }

        toString(resint, res);
        return res
    }
private:

    // 将路径转换为对应的字符串
    void toString(vector<vector<int>>& resint, vector<string>& res)
    {
        stringstream ss;
        for (int i = 0; i < resint.size(); i++)
        {
            string str = "";
            for (int j = 0; j < resint[i].size() - 1; j++)
            {
                ss.clear();
                string temp;
                ss << resint[i][j];
                ss >> temp;
                str += temp;
                str += "->";
            }
            
            ss.clear();
            string temp;
            ss << resint[i].back();
            ss >> temp;
            str += temp;

            res.push_back(str);
        }
    }
};